### DISCUSSION TITLE
MAIN.PY Runs With CTL-D but Not Power Off/On

### DISCUSSION THREAD

**Discussion by DS256** [micropython/micropython#10884] [RP2040 / Pico] [answered]
I am testing the passing of a GLOBAL variable between BOOT and MAIN. The programs are:

> #BOOT
> print("boot")
> with open("test.txt","w") as f:
>        f.write("BOOT.PY\n")
> global crash
> crash=True
> with open("test.txt","a") as f:
>      f.write("in Boot Crash="+str(crash)+"\n")

and

> #MAIN
> print("main")
> with open("test.txt","a") as f:
>     f.write("MAIN.PY\n")
> global crash
> if 'crash' in globals():
>     txt="In main Crash="+str(crash)
> else:
>     txt="In main NO CRASH"
> with open("test.txt","a") as f:
>     f.write(txt+"\n")
> print('done')

When I issue a CTL-D I get

> MPY: soft reboot
> boot
> main
> done
> MicroPython 699477d on 2023-01-16; Raspberry Pi Pico W with RP2040
> Type "help()" for more information.

TEST.TXT shows

_BOOT.PY
in Boot Crash=True
MAIN.PY
In main Crash=True_

When I disconnect and reconnect the USB power, all I see in TEST.TXT is

_BOOT.PY
in Boot Crash=True_

What am I doing wrong? I want MAIN.PY to run all the time

**Comment by davefes** [comment]
Try the same two tests with nothing in boot.py and only a `print('success') in main.py

The other thing to be aware of different things happen doing a CTRL-C, machine.reset() or a poweron reset (enable or boot buttons).  I am not familiar with the Pico.  Are you using an IDE?

**Comment by DS256** [comment]
As per @davefes suggestion I simplified my BOOT and MAIN and did some additional testing. I discovered the issues were:

- My BOOT program always created a new version of the log file. When I changed that to always append, I saw where MAIN was running..
- I was observing this from Thonny which seems to issue a SOFT_RESET when it connects to the PICO. This causes BOOT to run or maybe its Thonny running BOOT. In any case, MAIN is not run during a Thonny connection, only BOOT.

Here's what I found:

- POWER ON - cause=1; BOOT runs; MAIN runs
- RESET - cause=3; BOOT runs; MAIN runs
- SOFTREST - cause=3; BOOT runs; MAIN DOES NOT RUN

The RESET & SOFT_RESET machine calls were issued from the Micropython Interpreter shell in Thonny for the test.

According to HELP(MACHINE) 1 is PWRON_RESET; 3 is WDT_RESET. Not sure why 3 is the reason for both REST and SOFT_RESET on the PICO.

Still confused why RESET and SOFT_RESET return the same reset cause and one runs main and the other doesn't. May be when Thonny takes/relinquishes control. Still confusing to me.

**Comment by davefes** [comment]
I vaguely recall some recent threads on Thonny and Pico.  Have you checked issues on the Thonny website?

**Reply by DS256** [reply]
No but I will thanks.

It gets complicated since there are 3 related forums

- This MP forum
- MP on Raspberry PI Forums
- Thonny

**Reply by karfas** [reply]
> It gets complicated since there are 3 related forums

Maybe you should try to isolate the problem. 
If you can reproduce this issue using only putty or rshell or mpremote, it's a case for this forum.
Otherwise, thonny does things I don't want to know.

**Reply by DS256** [reply]
Thanks @karfas. I only found out about rshell yesterday and was unaware of putty and mpremote. Good suggestions.

**Comment by DS256** [comment, marked as answer]
@karfas put me onto trying to isolate the problem, so I took Thonny out of the equation. I set BOOT and MAIN to run autonomously, from a power cycle, through reset and soft_reset by passing the next reset type in a file.

{I had to replace indent spaces with '.' for readability}

> #BOOT
> import machine
> with open("reset.txt","r") as f:
> ....r=f.read()
> c = machine.reset_cause()
> with open("test.txt","a") as f:
> ....f.write("\n[boot] Started. Reset cause "+str(c)+" from last reset of "+r)

and

> #MAIN
> import machine
> with open("reset.txt","r") as f:
> ....r=f.read()
> with open("test.txt","a") as f:
> ....f.write("\n[main] Started. Last reset "+r)
> if r == "power":
> ....n="reset"
> elif r== "reset":
> ....n="soft"
> else:
> ....n="done"
> with open("test.txt","a") as f:
> ....f.write("\n[main] Next reset is "+n)
> with open("reset.txt","w") as f:
> ....f.write(n)
> if n == "reset":
> ....machine.reset()
> elif n == "soft":
> ....machine.soft_reset()

The resulting 'test.txt shows that the PICO W running Micropython standalone, without a connection to Thonny runs properly.

> [boot] Started. Reset cause 1 from last reset of power
> [main] Started. Last reset power
> [main] Next reset is reset
> [boot] Started. Reset cause 3 from last reset of reset
> [main] Started. Last reset reset
> [main] Next reset is soft
> [boot] Started. Reset cause 3 from last reset of soft
> [main] Started. Last reset soft
> [main] Next reset is done

I have a post open in the Thonny Forum I will link to this.

Thanks all for the guidance.

**Reply by scruss** [reply]
note for readability: put three backticks (_```_) on a line before and after your code and the indentation will be preserved

**Reply by DS256** [reply]
One additional point that was shared on the Raspberry PI Micropython Forum is that the PICO calls the watchdog-reboot function for a reset request which explains why a '3' for WDT is returned

**Reply by davehylands** [reply]
Rshell, thonny, mpremote, pyboard.py and probably others do a soft reset from the raw repl, and when that happens running main.py is suppressed. This is so that those programs can regain control. You can verify this by entering the raw repl and doing Control-D for the soft reset.
